File: code/Project_3/BT2/partition_comparison.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// Sinh tất cả phân hoạch của n
vector<vector<int>> generate_all_partitions(int n, int max_val = -1) {
    vector<vector<int>> result;
    
    if (max_val == -1) {
        max_val = n;
    }
    
    if (n == 0) {
        result.push_back({});
        return result;
    }
    
    for (int first = min(n, max_val); first >= 1; first--) {
        vector<vector<int>> sub_partitions = generate_all_partitions(n - first, first);
        for (auto& partition : sub_partitions) {
            vector<int> new_partition = {first};
            new_partition.insert(new_partition.end(), partition.begin(), partition.end());
            result.push_back(new_partition);
        }
    }
    
    return result;
}

// Đếm số phân hoạch có phần tử lớn nhất là k
int count_partitions_with_max(int n, int k) {
    if (k > n || k <= 0) {
        return 0;
    }
    
    vector<vector<int>> partitions = generate_all_partitions(n);
    int count = 0;
    
    for (const auto& partition : partitions) {
        if (!partition.empty() && partition[0] == k) {
            count++;
        }
    }
    
    return count;
}

// Sinh phân hoạch thành k phần
vector<vector<int>> generate_partitions_k_parts(int n, int k, int max_val = -1) {
    vector<vector<int>> result;
    
    if (max_val == -1) {
        max_val = n;
    }
    
    if (k == 1) {
        if (n >= 1 && n <= max_val) {
            result.push_back({n});
        }
        return result;
    }
    
    if (k > n || n <= 0) {
        return result;
    }
    
    int min_first = max(1, (n + k - 1) / k);
    int max_first = min(n - k + 1, max_val);
    
    for (int first = min_first; first <= max_first; first++) {
        vector<vector<int>> sub_partitions = generate_partitions_k_parts(n - first, k - 1, first);
        for (auto& partition : sub_partitions) {
            vector<int> new_partition = {first};
            new_partition.insert(new_partition.end(), partition.begin(), partition.end());
            result.push_back(new_partition);
        }
    }
    
    return result;
}

// Đếm số phân hoạch thành k phần
int count_partitions_k_parts(int n, int k) {
    vector<vector<int>> partitions = generate_partitions_k_parts(n, k);
    return partitions.size();
}

int main() {
    int n, k;
    
    cout << "Nhập n: ";
    cin >> n;
    cout << "Nhập k: ";
    cin >> k;
    
    if (n <= 0 || k <= 0) {
        cout << "n và k phải là số nguyên dương!\n";
        return 1;
    }
    
    // Tính p_max(n, k)
    int p_max = count_partitions_with_max(n, k);
    
    // Tính p_k(n)
    int p_k = count_partitions_k_parts(n, k);
    
    cout << "\nKết quả:\n";
    cout << "p_max(" << n << ", " << k << ") = " << p_max << "\n";
    cout << "p_" << k << "(" << n << ") = " << p_k << "\n";
    
    cout << "\nSo sánh:\n";
    if (p_max > p_k) {
        cout << "p_max(" << n << ", " << k << ") > p_" << k << "(" << n << ")\n";
    } else if (p_max < p_k) {
        cout << "p_max(" << n << ", " << k << ") < p_" << k << "(" << n << ")\n";
    } else {
        cout << "p_max(" << n << ", " << k << ") = p_" << k << "(" << n << ")\n";
    }
    
    return 0;
}

File: code/Project_3/BT2/partition_comparison.py
def generate_all_partitions(n, max_val=None):
    """Sinh tất cả phân hoạch của n"""
    if max_val is None:
        max_val = n
    
    if n == 0:
        return [[]]
    
    partitions = []
    for first in range(min(n, max_val), 0, -1):
        for partition in generate_all_partitions(n - first, first):
            partitions.append([first] + partition)
    
    return partitions

def count_partitions_with_max(n, k):
    """Đếm số phân hoạch của n có phần tử lớn nhất là k"""
    if k > n or k <= 0:
        return 0
    
    partitions = generate_all_partitions(n)
    count = 0
    
    for partition in partitions:
        if partition and partition[0] == k:
            count += 1
    
    return count

def generate_partitions_k_parts(n, k, max_val=None):
    """Sinh phân hoạch của n thành đúng k phần"""
    if max_val is None:
        max_val = n
    
    if k == 1:
        if 1 <= n <= max_val:
            return [[n]]
        else:
            return []
    
    if k > n or n <= 0:
        return []
    
    partitions = []
    min_first = max(1, (n + k - 1) // k)
    max_first = min(n - k + 1, max_val)
    
    for first in range(min_first, max_first + 1):
        remaining_partitions = generate_partitions_k_parts(n - first, k - 1, first)
        for partition in remaining_partitions:
            partitions.append([first] + partition)
    
    return partitions

def count_partitions_k_parts(n, k):
    """Đếm số phân hoạch của n thành đúng k phần"""
    partitions = generate_partitions_k_parts(n, k)
    return len(partitions)

def main():
    try:
        n = int(input("Nhập n: "))
        k = int(input("Nhập k: "))
        
        if n <= 0 or k <= 0:
            print("n và k phải là số nguyên dương!")
            return
        
        # Tính p_max(n, k)
        p_max = count_partitions_with_max(n, k)
        
        # Tính p_k(n)
        p_k = count_partitions_k_parts(n, k)
        
        print(f"\nKết quả:")
        print(f"p_max({n}, {k}) = {p_max}")
        print(f"p_{k}({n}) = {p_k}")
        
        print(f"\nSo sánh:")
        if p_max > p_k:
            print(f"p_max({n}, {k}) > p_{k}({n})")
        elif p_max < p_k:
            print(f"p_max({n}, {k}) < p_{k}({n})")
        else:
            print(f"p_max({n}, {k}) = p_{k}({n})")
            
        # Hiển thị chi tiết các phân hoạch
        print(f"\nChi tiết phân hoạch p_max({n}, {k}):")
        partitions_all = generate_all_partitions(n)
        count = 0
        for partition in partitions_all:
            if partition and partition[0] == k:
                print(f"  {partition}")
                count += 1
        if count == 0:
            print("  Không có phân hoạch nào")
            
        print(f"\nChi tiết phân hoạch p_{k}({n}):")
        partitions_k = generate_partitions_k_parts(n, k)
        if partitions_k:
            for partition in partitions_k:
                print(f"  {partition}")
        else:
            print("  Không có phân hoạch nào")
        
    except ValueError:
        print("Vui lòng nhập số nguyên hợp lệ!")

if __name__ == "__main__":
    main()